core/domain/collection_services.py

```diff
@@ -247,6 +247,77 @@ def is_collection_summary_editable(collection_summary, user_id=None):
         or collection_summary.community_owned)
 
 
+def get_learner_collection_dict_by_id(
+        collection_id, user_id, strict=True, version=None):
+    """Creates and returns a dictionary representation of a collection given by
+    the provided collection ID. This dictionary contains extra information
+    along with the dict returned by collection_domain.Collection.to_dict()
+    which includes useful data for the collection learner view. The information
+    includes progress in the collection, information about explorations
+    referenced within the collection, and a slightly nicer data structure for
+    frontend work.
+    """
+    collection = get_collection_by_id(
+        collection_id, strict=strict, version=version)
+
+    exp_ids = collection.exploration_ids
+    exp_summaries = (
+        exp_services.get_exploration_summaries_matching_ids(exp_ids))
+    exp_summaries_dict = {
+        exp_id: exp_summaries[ind] for (ind, exp_id) in enumerate(exp_ids)
+    }
+
+    # TODO(bhenning): Users should not be recommended explorations they
+    # have completed outside the context of a collection.
+    next_exploration_ids = None
+    completed_exploration_ids = None
+    if user_id:
+        completed_exploration_ids = get_completed_exploration_ids(
+            user_id, collection_id)
+        next_exploration_ids = collection.get_next_exploration_ids(
+            completed_exploration_ids)
+    else:
+        # If the user is not logged in or they have not completed any of
+        # the explorations yet within the context of this collection,
+        # recommend the initial explorations.
+        next_exploration_ids = collection.init_exploration_ids
+        completed_exploration_ids = []
+
+    collection_dict = collection.to_dict()
+    collection_dict['skills'] = collection.skills
+    collection_dict['next_exploration_ids'] = next_exploration_ids
+    collection_dict['completed_exploration_ids'] = (
+        completed_exploration_ids)
+    collection_dict['version'] = collection.version
+
+
+    # Insert an 'exploration' dict into each collection node, where the
+    # dict includes meta information about the exploration (ID and title).
+    for collection_node in collection_dict['nodes']:
+        summary = exp_summaries_dict.get(collection_node['exploration_id'])
+        if not summary:
+            raise utils.ValidationError(
```

Comment:
So, what's going to happen from the learner UX point of view in this case? Will they see a 500 error?

I'd have thought that it'd be better for them to see something telling them that the exploration has been deleted, but they should still be able to play the rest of the collection? Or is the point that this scenario will never occur (from the learner perspective)?